### ISSUE TITLE
Document use of escaped characters in selector-class-pattern

### ISSUE THREAD

**Issue by demrks** [stylelint/stylelint#2263]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

There seems to be a bug with the `selector-class-pattern` rule that produces an error although the regex works properly [in tests](http://regexr.com/3f2me)

> Which rule, if any, is this issue related to?

`selector-class-pattern`

> What CSS is needed to reproduce this issue?

e.g.

```css
.u-text-left { }
.u-size-11\/12\@sm { }
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "selector-class-pattern": "^.(o-|c-|u-|s-|is-|has-){1}([-]?[\\]?[/]?[a-z0-9]+){2,}(__[a-z0-9]([-]?[a-z0-9]+)*)?(--[a-z0-9-]+)?(\\@[a-z0-9-]+)?$"
  }
}
```

> Which version of stylelint are you using?

`7.7.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Webpack plugin

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

No, vanilla CSS

> What did you expect to happen?

No warnings to be flagged

> What actually happened (e.g. what warnings or errors you are getting)?

```shell
✖  Expected class selector ".u-text-left" to match specified pattern        selector-class-pattern
✖  Expected class selector ".u-size-11\/12\@sm" to match specified pattern   selector-class-pattern
```

**Comment by jeddy3** [comment]
@demrks Thanks for the report and for using the template.

You [shouldn't include](http://stylelint.io/user-guide/rules/selector-class-pattern/#options) the `.` in your regex. This change, however, only catches the `.u-text-left` selector. The `.u-size-11\/12\@sm` is still missed. I think this is a separate issue that needs further investigation.

The rule itself simply [uses `new RegExp(pattern)`], so it seems unlikely to be an issue here. I did notice that PostCSS [parses the selector](https://astexplorer.net/#/mJWRsDuLSx) to _include escaping_. Perhaps this has something to do with it?

@demrks Are you able to dig a little deeper and help us investigate this?

**Comment by demrks** [comment]
@jeddy3 Thanks so much, it works just fine if I change `"\\"` to `"\\\\"`:

```
^(o-|c-|u-|s-|is-|has-){1}([-]?[\\\\]?[/]?[a-z0-9]+){2,}(__[a-z0-9]([-]?[a-z0-9]+)*)?(--[a-z0-9-]+)?(\\\\@[a-z0-9-]+)?$
```

So, it's not a bug. Is this something that's worth mentioning in the docs?

**Comment by davidtheclark** [comment]
I think it's probably worth opening an issue in PostCSS and understanding the rationale. It might be a bug. Seems buggy to me, if the selector we get from the AST does not match the selector in the original stylesheet.

**Comment by demrks** [comment]
Ok, will do

**Comment by jeddy3** [comment]
@demrks Thanks for raising the issue.

Apparently [PostCSS is behaving as expected](https://github.com/postcss/postcss/issues/955#issuecomment-273034362). Therefore, do we either?:

1. Unescape the selector within the AST before doing any comparison/converting to regex.
2. Document this behaviour in the rule README.

**Comment by davidtheclark** [comment]
Unescaping seems to me the better plan.

**Comment by jeddy3** [comment]
Good stuff. Let's make sure this behaviour is consistent whenever we create regexs from user input. Milestoning as `8.0.0`.

**Comment by hudochenkov** [comment]
@jeddy3 @davidtheclark what do you mean by unescaping? I can't find a good solution to unescape strings. And it's seems can be a breaking operation.

Maybe we should escape a `string` containing RegExp before we do `new RegExp`. But if user has `RegExp` we wouldn't escape it. For escaping a string we could use `_. escapeRegExp`.

**Comment by davidtheclark** [comment]
@hudochenkov Feel free to try something out. If you can make it work, great! If not, we'll have to add this as a caveat.

**Comment by hudochenkov** [comment]
> Let's make sure this behaviour is consistent whenever we create regexs from user input. 

Seems like I didn't understand what this task about. Can someone explain what behaviour should be consistent? Examples won't hurt either :)

**Comment by davidtheclark** [comment]
@hudochenkov: I tried to create a test-case illustrating the confusion in #2520. If you can figure out how to fix it, or clarify it, that would be awesome.

**Comment by jeddy3** [comment]
Done in #2520
